nbconvert/preprocessors/execute.py

```diff
@@ -474,49 +502,77 @@ def _wait_for_reply(self, msg_id, cell=None):
                 self._check_alive()
                 cummulative_time += timeout_interval
                 if timeout and cummulative_time > timeout:
-                    self.log.error(
-                        "Timeout waiting for execute reply (%is)." % self.timeout)
-                    if self.interrupt_on_timeout:
-                        self.log.error("Interrupting kernel")
-                        self.km.interrupt_kernel()
-                        break
-                    else:
-                        raise TimeoutError("Cell execution timed out")
+                    self._handle_timeout()
+                    break
             else:
                 if msg['parent_header'].get('msg_id') == msg_id:
                     return msg
 
+    def _timeout_with_deadline(self, timeout, deadline):
+        if deadline is not None and deadline - monotonic() < timeout:
+            timeout = deadline - monotonic()
+
+        if timeout < 0:
+            timeout = 0
+
+        return timeout
+
+    def _passed_deadline(self, deadline):
+        if deadline is not None and deadline - monotonic() <= 0:
+            self._handle_timeout()
+            return True
+        return False
+
     def run_cell(self, cell, cell_index=0):
         parent_msg_id = self.kc.execute(cell.source)
         self.log.debug("Executing cell:\n%s", cell.source)
-        exec_reply = self._wait_for_reply(parent_msg_id, cell)
+        exec_timeout = self._get_timeout(cell)
+        deadline = None
+        if exec_timeout is not None:
+            deadline = monotonic() + exec_timeout
 
         cell.outputs = []
         self.clear_before_next_output = False
 
-        while True:
-            try:
-                # We've already waited for execute_reply, so all output
-                # should already be waiting. However, on slow networks, like
-                # in certain CI systems, waiting < 1 second might miss messages.
-                # So long as the kernel sends a status:idle message when it
-                # finishes, we won't actually have to wait this long, anyway.
-                msg = self.kc.iopub_channel.get_msg(timeout=self.iopub_timeout)
-            except Empty:
-                self.log.warning("Timeout waiting for IOPub output")
-                if self.raise_on_iopub_timeout:
-                    raise RuntimeError("Timeout waiting for IOPub output")
-                else:
-                    break
+        more_output = True
+        polling_exec_reply = True
+
+        while more_output or polling_exec_reply:
```

Comment:
`more_output` being used as a signaling mechanism feels like a pretty significant piece of internal logic. Could you add a comment explaining why we're taking this approach?

I would think a description about the intention of the separate roles of `more_output` and `polling_exec_reply` would be enough to address the curious individual.

Reply:
Great point, I added some detailed comments around this section to explain why it's written this way. Let me know if it's still not clear for future readers.